### ISSUE TITLE
Custom plugins are not executed unless global

### ISSUE THREAD

**Issue by ConaghFM** [Kong/kong#10929]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Kong version (`$ kong version`)

Kong 3.2.2

### Custom plugins are not executed unless they are configured as global plugins. 

I am working in kubernetes with KIC 2.9, but I do not think KIC is causing the problem. I have two custom plugins - one does some custom logging, the other is an auth plugin. Both exhibit the same behaviour. I have loaded the lua source as configmaps, and included in the KONG_PLUGINS variable. I can see the plugin is loaded correctly in the logs: 
`2023/04/24 10:13:11 [debug] 1#0: [lua] plugins.lua:284: load_plugin(): Loading plugin: my-custom-plugin`
If I configure the plugin using the KongClusterPlugin shown below, everything works as expected:
```
---
apiVersion: configuration.konghq.com/v1
kind: KongClusterPlugin
metadata:
  name: my-plugin-x
  namespace: kong
  annotations:
    kubernetes.io/ingress.class: kong
  labels:
    global: "true"
config:
  # some valid config
plugin: my-custom-plugin
```

However, I do not want this to be applied to every route. So I change to a non-global configuration:
```
---
apiVersion: configuration.konghq.com/v1
kind: KongClusterPlugin
metadata:
  name: my-plugin-x
  namespace: kong
  annotations:
    kubernetes.io/ingress.class: kong
  labels:
    global: "false"
config:
  # some valid config
plugin: my-custom-plugin
```
And apply the plugin annotation to the ingress [as described in the docs](https://docs.konghq.com/kubernetes-ingress-controller/latest/guides/using-kongplugin-resource/#associate-the-plugin-with-routing-configuration):
```
---
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: echo
  namespace: kong
  annotations:
    konghq.com/plugins: my-plugin-x
spec:
  ingressClassName: kong
  rules:
    - host: echo.company.com
      http:
        paths:
          - path: /
            pathType: Prefix
            backend:
              service:
                name: echo
                port:
                  number: 80
```
Then the plugin is not executed. The auth plugin no longer rejects invalid credentials, and the logging plugin stops writing logs. 
This also happens if I use a `KongPlugin` resource instead of `KongClusterPlugin`.
I can see the plugin applied to the route in the admin api:
```
https://localhost:51739/routes/6e971afb-522a-5612-aec1-06dad331bb13/plugins
{
  "next": null,
  "data": [
    {
      "config": "my-config",
      "tags": [
        "k8s-name:my-plugin-x",
        "k8s-namespace:kong",
        "k8s-kind:KongPlugin",
        "k8s-uid:5380dd90-ecbc-452e-bff0-97e956c7e79a",
        "k8s-group:configuration.konghq.com",
        "k8s-version:v1"
      ],
      "name": "my-custom-plugin",
      "service": null,
      "instance_name": null,
      "id": "2a7418a4-cc37-59a4-b02b-70d9f8d19314",
      "consumer": null,
      "created_at": 1684748703,
      "route": {
        "id": "6e971afb-522a-5612-aec1-06dad331bb13"
      },
      "enabled": true,
      "protocols": [
        "http",
        "https"
      ]
    }
  ]
}
```
That suggests KIC is doing it's job and has applied the config to Kong, and Kong knows it should execute the plugin, but is silently failing for some reason. 
I can't see any errors or anything else suspicious looking in the logs, even with debug enabled. 
I can use Kong's built-in plugins without any trouble, this only happens with my custom lua plugins.  

### Expected Behavior

Kong executes custom lua plugins when they are enabled on a service or route

### Steps To Reproduce

1. Deploy Kong 3.2 and KIC 2.9 in a Kubernetes 1.23 cluster
2. Load a custom Lua plugin into Kong
3. Apply the plugin to a service or route
4. Observe the plugin is loaded in the admin api, but has no effect

**Comment by StarlightIbuki** [comment]
Hi. Could you share the code of the plugins? Or at least disclose some information about them?

Possible reasons I can think of immediately:
1. The route is not chosen;
2. Some of the phase handlers does not run for route/service scoped plugins

**Comment by ConaghFM** [comment]
It turned out to be your second suggestion. Apparently the [rewrite phase is only executed for global plugins](https://docs.konghq.com/gateway/latest/plugin-development/custom-logic/#available-contexts). I checked the plugin code and sure enough, it was trying to execute in the rewrite function. I changed this to the access function, and it seems that solved the problem. The plugins are now executed on the routes they are applied to.
